### PR TITLE
feat: add Error extension trait for `Result<Option<Block>>`

### LINKED ISSUE #1374: Add Error extension trait used by RPC for `Result<Option<Block>>`
### Describe the feature

Since the provider return `Option` if a block does not exist we use code like this in rpc

```rust
self
    .client
    .block(_)
    .to_rpc_result()?
    .ok_or(EthApiError::UnknownBlockNumber)?;
```

ideally we have an extension trait that does this directly and does `interface::Result<Option<Block>>` to `RpcResult<Block>`

# TODO
* add an extension trait, perhaps `trait ToRpcResultExt {type Output; fn output_or_rpc_err(self) -> RpcResult<Self::Output>`
* impl for `interface::Result<Option<Block>>`

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by literallymarvellous**
Resolves: #1374 

Draft

### CODE REVIEW

**Comment by literallymarvellous** [comment]
@mattsse does this work?

File: crates/rpc/rpc/src/result.rs [hunk 1]
```diff
@@ -131,8 +132,38 @@ pub(crate) fn rpc_err(code: i32, msg: impl Into<String>, data: Option<&[u8]>) ->
     ))
 }
 
+pub(crate) trait ToRpcResultExt<T, Err> {
+    type Output;
+    type Error;
```
Comment (mattsse) [inline crates/rpc/rpc/src/result.rs:137]: we don't want this here.

this trait should be implemented manually for specific error types.

File: crates/rpc/rpc/src/result.rs [hunk 2]
```diff
@@ -131,8 +132,38 @@ pub(crate) fn rpc_err(code: i32, msg: impl Into<String>, data: Option<&[u8]>) ->
...
+impl<T, Err> ToRpcResultExt<T, Err> for RethResult<Option<T>>
+where
+    T: Clone,
+    Err: Error + Into<RpcError>,
+    Self: Sized,
+{
+    type Output = T;
+    type Error = Err;
+
+    fn output_or_rpc_err(self, err: Err) -> RpcResult<Self::Output> {
```
Comment (mattsse) [inline crates/rpc/rpc/src/result.rs:151]: this should not take err.

**Review by mattsse — CHANGES_REQUESTED** [review]

**Comment by gakonst** [comment]
@mattsse should we just make ToRpcResult do this always? doing `.to_rpc_result()?. output_or_rpc_err()?` doesnt look good. i could see calling it like `res()` or something instead, but would pref something less verbose

**Comment by mattsse** [comment]
> @mattsse should we just make ToRpcResult do this always? doing .to_rpc_result()?. output_or_rpc_err()?

if we want to go from something like `Result<Option<T>>` to `RpcResult<T>` then we need a helper, that's what `Result<Option<T>>::output_or_rpc_err` would be for. and it's just `output_or_rpc_err` which is intended to be a `map+transpose` but in one step.

we could however replace `.to_rpc_result()` with a From impl

**Comment by gakonst** [comment]
I guess it's fine, mostly nitpicking. Merging then.

**Review by mattsse — APPROVED** [review]
awesome, 

I#ve renamed the function to make it clearer what this actually does.

**Pull request merged**